### ISSUE TITLE
[TF2] [Bug] Flamethrower damage ramps-up notably slower at point-blank compared to any other range

### ISSUE THREAD

**Issue by VirtualSoda** [ValveSoftware/Source-1-Games#3631]
The Flamethrower ramps-up slower than intended at point-blank range. This goes against the trend of having damage ramp-up faster the younger the flame particles are. 

Video demonstration: https://www.youtube.com/watch?v=IhQ3N8F7XR0
You can clearly see that when I start firing at point-blank range (where you do a maximum of 13 damage) It ramps-up significantly slower compared to being at close-range (where you do a maximum of 12 damage).

This is a rather big issue because it significantly reduces Pyro's damage output, while not being intentional. Hopefully this can be addressed.

**Comment by VirtualSoda** [comment]
I want to add onto this; that for one reason or another, ramp-up seems to work with the speed inteded if you flick to your target. 
This is really weird.

https://www.youtube.com/watch?v=ksMzURCwrXw

Video shows that the damage ramps-up faster when you flick to your target before hitting them.

**Comment by VirtualSoda** [comment]
I want to add another thing. I found out something today that struck me as very odd:

In this video: https://youtu.be/e-9JdfNUH2o?t=53 you can see that the Flamethrower reaches maximum damage (13) very quickly. This is odd, because it usually takes ages to ramp-up to 13 damage at point-blank range.  I thought it could maybe be something I hadn't taken into account before. But I tried replicating it myself, to no succes. On top of this, you can see damage ramp-up starting at 7 damage in this video, but it starts from 6 in-game currently.

I wanted to look further into this, so I tried finding more videos about the Blue Moon flamethrower changes. I found a reddit thread with a comment: https://www.reddit.com/r/truetf2/comments/az5gih/comment/ei6zk99/?utm_source=share&utm_medium=web2x&context=3 and it included this gif 

https://user-images.githubusercontent.com/25712612/141989648-2f177c76-f4c0-4d9d-9ce2-b0e39c045208.mp4



As you can see in this GIF; ramp-up starts at 7, and reaches 13 very quickly. More important, it can reach 13 damage from further away than in-game currently. 

Somehow, this isn't how it works right now. Some update must have changed this. This GIF was dated March 3rd, 2019. So I think this update changed something: 
![5594-14db24fa-cdde-4f0f-a0de-6bc363eabf12](https://user-images.githubusercontent.com/25712612/141989800-b7321ccd-24c6-48fd-80bf-d0128cf5118f.png)

**Comment by VirtualSoda** [comment]
https://www.youtube.com/watch?v=6ewnf4ReiAQ

The Spy is a bot and the Heavy is a real player. Nothing varied about ramp-up with the Spy. Damage varied with the Heavy.

On top of this, you can clearly see in the video that I am doing more damage to the Spy that is further away from me than the Heavy. 

I am atleast slightly sure now that is is networking related.

**Comment by chaorace** [comment]
I did some research into this bug in the abovelinked duplicate issue. The root cause is now known and it considerably changes the scope of this issue, so I'm hoping the original issue owner can update the title/description to better reflect the underlying bug more accurately.

Summing it up below...

The "heat" increment is competing with a constant decay rate, so naturally the game designers had to tune the associated parameters such that `buildrate > decayrate`. When you look at the code, though... the buildrate is actually functionally less than the decayrate! The only reason it works at all is because the _actual_ in-game buildrate is **double** the value-as-programmed.

This leads to a pretty clear conclusion: there's a double-counting bug in the "heat" system and instead of fixing it the developers simply tuned the heat parameters around it. They essentially built around the bug, assuming it would behave consistently. Unfortunately, it does not behave consistently. There are [two well-known cases](https://www.youtube.com/watch?v=JqaI5LhNalk) where double-counting fails to trigger, both of which share a common root cause:
1. Shooting victims who are backed against an axis-aligned wall.
2. Shooting victims who are axis-aligned and 70hu or nearer to you.

In the first case, the flame particles are blocked from exiting the victim's AABB.
In the second case, the flame particles are spawned within the victim's AABB (_and are thus prevented from making initial entry_).
(_2nd case happens because the flamethrower muzzle projects out ~70hu, which is enough to overlap inside another player's AABB_)

In both cases, the issue only manifests if the victim is not moving. Moving causes the 2nd hit to register even when the other AABB conditions are being met. The no-movement requirement renders the 2nd case a mostly theoretical concern, since most players will refuse to hold still while you're actively shooting them. The 1st case is a lot more stubborn, though, since players backed into a corner often have no choice but to remain still.

**tl;dr**: the intended heat growth requires your victim to be moving OR for your flames to both _ENTER AND EXIT_ the victim's AABB. This is very unintuitive and the current state of the mechanic inadvertently (in the axis-aligned walls case) invisibly punishes players for playing the class as originally intended.

**Reproduction Steps** (_listen server OK, buddy required_):
1. `sv_cheats 1`
2. `developer 1`
3. `tf_debug_flamethrower  1`
4. `ent_absbox player`
5. `hud_damagemeter 1`
6. `hud_damagemeter_period 0.5`
7. `tf_weapon_criticals 0`
8. Buddy uses `hurtme -100000`
9. Position yourself so that the origin point of your spawning flames is inside buddy's AABB (_origin point is easy to see w/ yellow line_)
10. Shoot for 3s -- note how the damage meter does not go above 100
11. Back up until your flames are spawning JUST BARELY outside the buddy's AABB
12. Shoot for 3s -- note how the damage meter gets well above 100
13. Walk forward a tiny bit so that your spawning-in flames are JUST BARELY inside the buddy's AABB
14. Shoot for 3s -- note how the damage meter no longer goes above 100
